### PR TITLE
Nijdam/ch1808/add enacl support

### PR DESCRIPTION
**Pull request by madninja**
This adds support for mulitple keytypes for crypto:

* Add support for ed25519, while keeping support for ecc_compact
* Rename address to pubkey_bin to make it clearer what the thing actually is

### CODE REVIEW

File: src/libp2p_crypto.erl [hunk 1]
```diff
@@ -2,119 +2,155 @@
...
 
--spec to_pem(private_key()) -> binary().
-to_pem(PrivKey) ->
+-spec keys_to_bin(key_map()) -> binary().
+keys_to_bin(#{secret := {ecc_compact, PrivKey}, public := {ecc_compact, _PubKey}}) ->
     PemEntry = public_key:pem_entry_encode('ECPrivateKey', PrivKey),
-    public_key:pem_encode([PemEntry]).
+    <<?KEYTYPE_ECC_COMPACT:8, (public_key:pem_encode([PemEntry]))/binary>>;
+keys_to_bin(#{secret := {ed25519, PrivKey}, public := {ed25519, PubKey}}) ->
+    io:format("PUB ~p PRIV: ~p", [byte_size(PubKey), byte_size(PrivKey)]),
```
Comment (Vagabond) [inline src/libp2p_crypto.erl]: Remove?

File: src/libp2p_crypto.erl [hunk 2]
```diff
@@ -2,119 +2,155 @@
...
+    Bin = keys_to_bin(KeysMap),
+    file:write_file(FileName, Bin).
 
--spec to_pem(private_key()) -> binary().
-to_pem(PrivKey) ->
+-spec keys_to_bin(key_map()) -> binary().
+keys_to_bin(#{secret := {ecc_compact, PrivKey}, public := {ecc_compact, _PubKey}}) ->
     PemEntry = public_key:pem_entry_encode('ECPrivateKey', PrivKey),
-    public_key:pem_encode([PemEntry]).
+    <<?KEYTYPE_ECC_COMPACT:8, (public_key:pem_encode([PemEntry]))/binary>>;
```
Comment (Vagabond) [inline src/libp2p_crypto.erl]: You're mixing binary and ascii data here, this seems inadviseable?

Comment (Vagabond) [reply src/libp2p_crypto.erl]: Maybe it's better to just do a binary format similar to the 25519 curve?

File: src/libp2p_crypto.erl [hunk 3]
```diff
@@ -2,119 +2,155 @@
...
+        _ -> false
+    end.
+
 
--spec address_to_b58(address()) -> string().
-address_to_b58(Addr) ->
-    address_to_b58(16#00, Addr).
+-spec bin_to_b58(binary()) -> string().
+bin_to_b58(Bin) ->
+    bin_to_b58(16#00, Bin).
```
Comment (Vagabond) [inline src/libp2p_crypto.erl]: is 16#00 the 'version'?

File: test/crypto_SUITE.erl [hunk 4]
```diff
@@ -1,71 +0,0 @@
--module(crypto_SUITE).
```
Comment (Vagabond) [inline test/crypto_SUITE.erl]: Why was this removed?

File: src/libp2p_crypto.erl (hunk 3 above)
Comment (Vagabond) [reply src/libp2p_crypto.erl]: Wondering why the 'version' is not just the key type?

File: src/libp2p_crypto.erl (hunk 2 above)
Comment (madninja) [reply src/libp2p_crypto.erl]: but isn't `pem_encode(PemEntries) -> binary()` the right type signature for creating a binary here?

File: test/crypto_SUITE.erl (hunk 4 above)
Comment (madninja) [reply test/crypto_SUITE.erl]: Because the SUITE was moved into the unit tests inside the .erl file a while ago. We just never noticed

File: src/libp2p_crypto.erl (hunk 2 above)
Comment (madninja) [reply src/libp2p_crypto.erl]: Resolved with a87effe

File: src/libp2p_crypto.erl (hunk 3 above)
Comment (madninja) [reply src/libp2p_crypto.erl]: Mostly to avoid the types and versions from overlapping.

File: .travis.yml [hunk 5]
```diff
@@ -13,6 +13,7 @@ before_install:
   - cd libsodium && sudo ./configure && sudo make check && sudo make install && cd ..
 
 cache:
+  ccache: true
```
Comment (benoitc) [inline .travis.yml]: for what do you need it? is this to speed the build of rocksdb? there is a branch that speed it. Also eventually ccache can be added to it. Let me know :)

**Comment by madninja** [comment]
Yeah it’s to speed up rocksdb and libsodium builds. You have something that helps?

**Comment by benoitc** [comment]
using the `static_build` branch you can run a // build using a version of cmake that supports it: 

```
ERLANG_ROCKSDB_OPTS="-DWITH_LZ4=ON -DWITH_SNAPPY=ON" ERLANG_ROCKSDB_BUILDOPTS="-j" rebar3 compile
```

will run the build in parralel with all cores with an installed version of snappy and lz4. I can add the support of `ccache` too if needed. You can also use a packaged version of rocksdb (if 0.17.2) for it using the option `WITH_SYSTEM_ROCKSDB=ON`. Let me know if it improves the build for you.

**Comment by benoitc** [comment]
also with TRAVIS you can cache the built of erlang-rocksdb. This is what most do as of today.

**Comment by madninja** [comment]
Right that's what I'm trying to do. The apt source list for travis is
pretty minimal so I'm not sure if rocksdb is even in there at all today,
let alone on a recent version

**Comment by madninja** [comment]
@benoitc I'm going to back this out and just try to speed up libsodium builds. We can tackle the rocks speedup separately.

**Comment by benoitc** [comment]
@madninja ok. fyi https://gitlab.com/barrel-db/erlang-rocksdb/commit/7cf8d66b5f1b0458891b7bdb6e1275f471db843c has support for ccache also

**Comment by madninja** [comment]
Oh very nice! Can’t wait for this to merge up!

**Comment by benoitc** [comment]
@madninja https://hex.pm/packages/rocksdb/1.0.0 & https://gitlab.com/barrel-db/erlang-rocksdb/blob/master/doc/customize_rocksdb_build.md ;)
